js/apiService.js

```diff
@@ -0,0 +1,29 @@
+class APIService {
+    constructor() {
+        if (!APIService.instance) {
+            this.cache = {};
+            APIService.instance = this;
+        }
+        return APIService.instance;
+    }
+
+    async fetchPosts(page = 1, limit = 10, search = '') {
+
+        // to maintain a cache to prevent api calling for already fetched data
+        const cacheKey = `${page}-${limit}-${search}`;
+        if (this.cache[cacheKey]) {
+            return this.cache[cacheKey];
+        }
+
+        // api call
+        const response = await fetch(`https://dummyjson.com/posts?skip=${(page - 1) * limit}&limit=${limit}&search=${search}`);
```

Comment:
create another class, BaseAPIService ( it have all method, get, put, post ) already explained https://teams.microsoft.com/l/message/19:b6b5afdaf8654b349738eed1f5345401@thread.v2/1721645452471?context=%7B%22contextType%22%3A%22chat%22%7D

Reply:
done.